[404 server] Add prometheus metrics

cmluciano opened the pull request:
Metrics tracked:
- total number of requests
- duration of request
- status codes and method

I am representing duration for the request in milliseconds because my
test request almost always returned way under 1 second.

Commit:
Catch error from listenandserve

This change catches any potential errors from http.listenandserve
and ouputs to stderr.

Commit:
Add prometheus metrics

Metrics tracked:
- total number of requests
- duration of request

I am representing duration for the request in milliseconds because my
test request almost always returned way under 1 second.

Closes kubernetes/ingress#237

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/2376)
<!-- Reviewable:end -->

cmluciano commented:
I've tried adding the prometheus library using godep, but the workspace directory keeps overriding it. I tried removing it since it will be unsupported in 1.8, but the build still fails while trying to locate prometheus.

luxas commented:
cc @bprashanth

cmluciano commented:
Any advice on the Godep pull @bprashanth ?

piosz commented:
Try do the following:
```
rm -rf Godeps/_workspace
go get github.com/prometheus/client_golang/prometheus
go get github.com/prometheus/client_golang/prometheus/promhttp
godep save ./...
```

cmluciano commented:
Not sure if it related to the AWS stuff, but I cannot retrieve the build log for some reason

cmluciano commented:
Can i get a retest on this @piosz ? The Travis build seems to have exited prematurely.

piosz commented:
Closed/reopened in order to retrigger tests.

piosz commented:
Also you need to include `vendor` directory into your PR.

cmluciano commented:
Fingers crossed :)

piosz commented on 404-server/Godeps/Godeps.json:
```diff
@@ -1,8 +1,15 @@
+	"GodepVersion": "v78",
 	"Packages": [
-		"./..."
+		"./.."
```
You've made a mistake here. You should type:
```
godep save ./...
```
but you've typed
```
godep save ./..
```
The different is just one dot. The should be 3. Something weird had happened. Please try again.

cmluciano commented:
Green Go the Rushes O !

Nice catch on the missing . @piosz

piosz commented:
LGTM from me
@bowei can you please take a look or delegate?
@mikedanese can you please approve?

piosz commented:
@cmluciano thanks for the PR

bowei commented on 404-server/server.go:
```diff
@@ -36,4 +37,10 @@ func main() {
+	// TODO: Use .Shutdown from Go 1.8
+	err := http.ListenAndServe(fmt.Sprintf(":%d", *port), nil)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "could not start http server: %s\n", err)
```
log.Fatal?

bowei commented on 404-server/server.go:
```diff
@@ -22,21 +22,45 @@ import (
 		fmt.Fprint(w, "default backend - 404")
+
+		duration := time.Now().Sub(start).Seconds() * 1e3
+		fmt.Println(duration)
```
did you mean to print this to stdout?

cmluciano replied:
Oops, that was for testing the time units, removing

bowei commented on 404-server/metrics.go:
```diff
@@ -0,0 +1,52 @@
+		Subsystem: subsystem,
+		Name:      "request_duration_milliseconds",
+		Help:      "Histogram of the time (in milliseconds) each request took.",
+		Buckets:   append([]float64{.001, .003}, prometheus.DefBuckets...),
```
[minor] wonder if recording at microsecond resolution is useful?

cmluciano replied:
I did have this at microseconds, but I didn't know what the average response time would be "in the wild". Would you like me to switch to microseconds?

bowei commented on 404-server/metrics.go:
```diff
@@ -0,0 +1,52 @@
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "response_status_count_total",
+		Help:      "Counter of response status codes.",
```
isn't this always 404?

bowei reviewed (commented):
some nits

cmluciano commented:
I removed the request code part since it should always be a 404. I also removed the testing println that I missed. 

@bowei LMK what you think of https://github.com/kubernetes/contrib/pull/2376#discussion_r104457237

bowei commented:
/approve /lgtm

mikedanese commented:
/approve
